src/Features/Core/Portable/CodeRefactorings/SyncNamespace/AbstractSyncNamespaceService.ChangeNamespaceCodeAction.cs

```diff
@@ -0,0 +1,526 @@
+            }
+
+            private ImmutableArray<string> CreateAllContainingNamespaces(ImmutableArray<string> parts)
+            {
+                var builder = ArrayBuilder<string>.GetInstance();
+                for (var i = 1; i <= OldNamespaceParts.Length; ++i)
+                {
+                    builder.Add(string.Join(".", OldNamespaceParts.Take(i)));
+                }
+
+                return builder.ToImmutableAndFree();
+            }
+
+            private ImmutableArray<SyntaxNode> CreateImports(Document document, ImmutableArray<string> names, bool withFormatterAnnotation)
+            {
+                var generator = SyntaxGenerator.GetGenerator(document);
+                var builder = ArrayBuilder<SyntaxNode>.GetInstance(names.Length);
+                for (var i = 0; i < names.Length; ++i)
+                {
+                    builder.Add(CreateImport(generator, names[i], withFormatterAnnotation));
+                }
+
+                return builder.ToImmutableAndFree();
+            }
+
+            private SyntaxNode CreateImport(SyntaxGenerator syntaxGenerator, string name, bool withFormatterAnnotation)
+            {
+                var import = syntaxGenerator.NamespaceImportDeclaration(name);
+                if (withFormatterAnnotation)
+                {
+                    import = import.WithAdditionalAnnotations(Formatter.Annotation);
+                }
+                return import;
+            }
+
+            private async Task<(Solution, ImmutableArray<DocumentId>)> ChangeNamespaceToMatchFoldersAsync(Solution solution, DocumentId id, CancellationToken cancellationToken)
+            {
+                var document = solution.GetDocument(id);
+                var targetNamespace = _state.TargetNamespace;
+
+                var declarationRoot = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+                var container = declarationRoot.DescendantNodes().FirstOrDefault(node => node is TNamespaceDeclarationSyntax) ?? declarationRoot;
+
+                var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+                var declaredSymbols = GetDeclaredSymbolsInContainer(semanticModel, container, cancellationToken);
+
+                var documentSet = ImmutableHashSet.Create(document);
+                var editor = await DocumentEditor.CreateAsync(document, cancellationToken).ConfigureAwait(false);
+
+                var refLocationsInCurrentDocument = new List<ReferenceLocation>();
+                var refLocationsInOtherDocuments = new List<ReferenceLocation>();
+
+                foreach (var declaredSymbol in declaredSymbols)
+                {
+                    cancellationToken.ThrowIfCancellationRequested();
+
+                    var refSymbols = await SymbolFinder.FindReferencesAsync(
+                      declaredSymbol,
+                      solution,
```

Comment:
you should call the overload that takes SymbolAndProjectId.  It will be much faster and can run out of proc.  Also, we should do this work in parallel since there may be many symbols we need to find and fixup.

Reply:
Done.